### ISSUE TITLE
CircularBuffer should implement withContiguousStorageIfAvailable

### ISSUE THREAD

**Issue by Lukasa** [apple/swift-nio#1215]
Some algorithms on Sequences can implement a fast path in cases where that Sequence can vend an interior pointer to a contiguous buffer, by way of [`withContiguousStorageIfAvailable`](https://developer.apple.com/documentation/swift/sequence/3128824-withcontiguousstorageifavailable). This includes SwiftNIO, which has a fast-path for writing generic bytes objects that can vend such a pointer.

Sadly, our very-own CircularBuffer is not one of those data types. But it could be! While it is _sometimes_ split into two regions, it is also sometimes the case that the data is actually contiguous. For that reason, we should implement this method.

The implementation is straightforward:

- When the buffer is a single contiguous region, we should vend the interior pointer.
- When the buffer is multiple regions, we should not.

**Comment by Keno42** [comment]
I’ll work on this.

**Comment by Keno42** [comment]
Hi @Lukasa ,

I took a look at related code and faced a problem.

I thought it would be ideal if `withContiguousStorageIfAvailable(_:)` could pass the storage's pointer that points `_buffer[self.headBackingIndex]` if possible, but because the default declaration of the method requires the pointer has a generic type `UnsafeBufferPointer <Self.Element>` while the buffer has a generic type `ContiguousArray<Element?>`, it seemed impossible to do so. In case `Element` is `Int`, `MemoryLayout` tells sizes of `Int` and `Int?` are 8 and 9, respectively. 

Here's my first question:
1. Do I understand the situation correct?

If then, how should it be? Here are 2 options I came up with.
2-a. Make a different function which works with `UnsafeBufferPointer<Element?>`.
2-b. Flatmap `Element?` to `Element` and use it. (I guess it costs too much considering the original purpose)

I would appreciate hearing your thoughts on this.
Thanks.
